Allow to mix static and dynamic tools

mariofusco opened the pull request.

mariofusco committed:
Allow to mix static and dynamic tools

dliubarskyi commented on langchain4j/src/main/java/dev/langchain4j/service/tool/ToolService.java:
```diff
@@ -88,26 +72,19 @@ public void tools(Collection<Object> objectsWithTools) {
-        List<ToolSpecification> toolsSpecs = new ArrayList<>();
-        Map<String, ToolExecutor> toolExecs = new HashMap<>();
+        List<ToolSpecification> toolsSpecs = new ArrayList<>(this.toolSpecifications);
+        Map<String, ToolExecutor> toolExecs = new HashMap<>(this.toolExecutors);
```
Could you please add a validation that there are no duplicates in tool names between static and dynamic tools?

mariofusco replied:
Good point, but what should we do in case of a duplication? I'd say that the static one should take precedence, but I agree that this is a bit arbitrary. Any other suggestion?

dliubarskyi replied:
When there is a conflict between 2 static tools, we just fail with exception (see `ToolService.tools`). I guess we should do the same here?

mariofusco replied:
> When there is a conflict between 2 static tools, we just fail with exception (see `ToolService.tools`). I guess we should do the same here?

Done. Please check again.

dliubarskyi reviewed (commented):
@mariofusco thank you!

Could you please update the documentation? I guess https://docs.langchain4j.dev/tutorials/tools#specifying-tools-dynamically section scould specify the behaviour when both static and dynamic tools are configured.

It would also be great to keep the checklist from the PR template as it helps to not forget/oversee inportant points.

mariofusco committed:
check duplicated tools + add docs

mariofusco commented:
I added a check for duplicated tools, always giving precedence to statically defined ones, and added another test to verify this behavior.  I also added a sentence in the docs to clarify this.

@dliubarskyi please give this another look.

mariofusco committed:
fail on duplicated tools

dliubarskyi reviewed (approved):
@mariofusco thank you!
